src/btree/bt_cursor.c

```diff
@@ -1231,10 +1232,19 @@ __btcur_update(WT_CURSOR_BTREE *cbt, WT_ITEM *value, u_int modify_type)
     /* If our caller configures for a local search and we have a page pinned, do that search. */
     if (F_ISSET(cursor, WT_CURSTD_UPDATE_LOCAL) && __cursor_page_pinned(cbt, true)) {
         __wt_txn_cursor_op(session);
+        WT_ERR(__wt_txn_autocommit_check(session));
 
         WT_ERR(btree->type == BTREE_ROW ? __cursor_row_search(cbt, true, cbt->ref, &leaf_found) :
                                           __cursor_col_search(cbt, cbt->ref, &leaf_found));
-        if (leaf_found)
+        /*
+         * If searching an already pinned page doesn't find an exact match and returns the first or
+         * last page slots, discard the results and search the full tree, a non-existent record
+         * might belong on an entirely different page. This test is simplistic as we're ignoring
```

Comment:
I find both of these (identical) comments confusing. It is essentially opposite of what the code directly below it is actually doing. They talk about not finding an exact match and discarding the results, searching the full tree, etc, but the code below is actually testing for an exact match and returning those results.

Reply:
Reworked in a21f9e0e7.